test/cpp/jit/test_module_api.cpp

```diff
@@ -97,5 +97,50 @@ void testModuleParameter() {
   ASSERT_TRUE(m.hasattr("none_param2"));
 }
 
+void testModuleDeepcopy() {
+  auto cu = std::make_shared<CompilationUnit>();
+  auto cls = ClassType::create("foo.bar", cu, true);
+  auto int_attr = "int_attr";
+  auto tensor_attr = "tensor_attr";
+  cls->addAttribute(int_attr, IntType::get());
+  cls->addAttribute(tensor_attr, TensorType::get());
+  Module m(cu, cls);
+  m.setattr(int_attr, IValue(2));
+  m.setattr(tensor_attr, at::randn(5));
+
+  Module m2 = m.deepcopy();
+  Module m3 = m.clone_instance();
+  // Make sure copy works
+  ASSERT_EQ(m2.attr(int_attr).toInt(), 2);
+  ASSERT_EQ(m3.attr(int_attr).toInt(), 2);
+
+  // Both deepcopy and clone_instance will preserve the type
+  ASSERT_EQ(m.type(), m2.type());
+  ASSERT_EQ(m.type(), m3.type());
+
+  // change int value of copied instances
+  m2.setattr(int_attr, IValue(3));
```

Comment:
Can you add tests to verify aliasing logic? 
Module 
   self.a =  [t1, t2]
   self.b = (t3, t1)

verify that a and b overlap for example in the deepcopy

Reply:
In general, having more tests to verify edge cases is good?  The key functionality of deepcopy 
1) original and deepcopy cannot have side effect
2) Aliasing is preserved in deep copy.

Please also add a tests for string